worlds/hk/__init__.py

```diff
@@ -514,23 +514,40 @@ def create_vanilla_location(self, location: str, item: Item):
     def collect(self, state, item: HKItem) -> bool:
         change = super(HKWorld, self).collect(state, item)
         if change:
+            prog_items = state.prog_items
             for effect_name, effect_value in item_effects.get(item.name, {}).items():
-                state.prog_items[effect_name, item.player] += effect_value
-        if item.name in {"Left_Mothwing_Cloak", "Right_Mothwing_Cloak"}:
-            if state.prog_items.get(('RIGHTDASH', item.player), 0) and \
-                    state.prog_items.get(('LEFTDASH', item.player), 0):
-                (state.prog_items["RIGHTDASH", item.player], state.prog_items["LEFTDASH", item.player]) = \
-                    ([max(state.prog_items["RIGHTDASH", item.player], state.prog_items["LEFTDASH", item.player])] * 2)
+                prog_items[effect_name, self.player] += effect_value
+
+            # a directional overflow in dash grants an upgrade for the other side
+            if item.name == "Right_Mothwing_Cloak" and \
+                    prog_items["RIGHTDASH", self.player] > 2 > prog_items["LEFTDASH", self.player]:
```

Comment:
I don't get why it's checking for greater than 2?

Reply:
Because, the way I understood it, is that when that direction can't be upgraded anymore it upgrades the other instead. And there's 2 upgrades per direction.

Reply:
Yeah, I don't think this implementation is quite correct

So, deep diving into ItemChanger we can find this implementation for split cloak: https://github.com/homothetyhk/HollowKnight.ItemChanger/blob/master/ItemChanger/Resources/items.json#L5582
https://github.com/homothetyhk/HollowKnight.ItemChanger/blob/master/ItemChanger/Tags/ItemTreeTag.cs

This means that if left cloak is unobtained, you get one left dash. If left cloak has already been obtained, it will be shade cloak (either left or full depending if you have right dash or not). After that, the item tree is exhausted and you would actually get nothing. The same is true for right dash. Since AP makes 2 left and 2 right, I believe it should be the case that precisely the 4th dash you obtain would just do nothing.

Tldr I think instead of overflowing you'd just no-op in such cases. But I'd need to verify the behavior is as I described above in an actual AP game

Reply:
I know that if you make an HK AP game, you end up with 3 Mothwing Cloaks; 2 in one direction, 1 in the other. I'm not sure why/how it makes 3 instead of 4, but there is [code to convert the Shade Cloak randomly into a Left or Right Mothwing Cloak](https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/hk/__init__.py#L294-L298) based on your settings

Reply:
Oh, I actually read that code and i completely missed the if/randint (so in a normal game, without usage of getitem, there should only be 3)

So actually I think this overflowing logic just doesn't need to exist at all

Reply:
Ok, I finally remembered to go double check on this. here is what I found

* 3 dashes get placed in current AP. This is all well and good. One left, one right, one which is either left or right at random
* The correct handling should be something like the following:
  ```python
  has_left_dash = prog_items["LEFTDASH", self.player]
  has_right_dash = prog_items["RIGHTDASH", self.player]
  has_any_shade_cloak = prog_items["LEFTDASH", self.player] > 1 or prog_items["RIGHTDASH", self.player] > 1
  if item.name == "Left_Mothwing_Cloak":
    # we always get 1 leftdash - then we'll go see if we need any special handling to upgrade
    # to shade cloak
    prog_items["LEFTDASH", self.player] += 1
    if has_left_dash and has_right_dash:
      # we already had both directional dashes, so we want right shade cloak too
      prog_items["RIGHTDASH", self.player] += 1  
    elif has_any_shade_cloak:
      # we already had right shade cloak so make this our left shade cloak too
      prog_items["LEFTDASH", self.player] += 1
  ```

In the code above, I've not included any handling for dupes because they don't exist. Mirroring this logic for right dash and inverting the logic for reverse fill (which I understand is the goal of this PR) is left as an exercise for the reader